### PR TITLE
Allow to save and import HARs (HTTP archives)

### PR DESCRIPTION
**Pull request by bunysae**
Fixes issue zaproxy/zaproxy#521 and replaces PR #2406.

### CODE REVIEW

**Comment by kingthorin** [comment]
@bunysae we've decided the code here is complete as far as the original issue goes so we're going to close the issue and payout the bounty. However, this PR will stay open and be re-worked slightly. (Merging the code into an existing import addon-on,and making minor tweaks.)

File: addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarApi.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.logging.log4j.Logger;
+import org.zaproxy.zap.extension.api.ApiAction;
+import org.zaproxy.zap.extension.api.ApiException;
+import org.zaproxy.zap.extension.api.ApiException.Type;
+import org.zaproxy.zap.extension.api.ApiImplementor;
+import org.zaproxy.zap.extension.api.ApiResponse;
+import org.zaproxy.zap.extension.api.ApiResponseElement;
+import org.zaproxy.zap.utils.ApiUtils;
+
+/** The API for importing URLs from a file. */
```
Comment (thc202) [inline addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarApi.java:34]: `URLs` → `data`? It allows to import whole messages.

File: addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarApi.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+public class HarApi extends ApiImplementor {
+
+    private static final Logger LOG = LogManager.getLogger(HarApi.class);
+    private static final String PREFIX = "importexport";
```
Comment (thc202) [inline addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarApi.java:38]: I'd suggest `exim` (the same as the add-on ID), to reduce the likelihood of conflicts with other add-ons.

File: addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarApi.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    private static final String ACTION_IMPORTHAR = "importhar";
+    private static final String PARAM_FILE_PATH = "filePath";
+
+    /** Provided only for API client generator usage. */
```
Comment (thc202) [inline addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarApi.java:42]: Can be removed.

File: addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarImporter.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                return;
+            }
+        } else {
+            historyRef = message.getHistoryRef();
+        }
+
+        ExtensionHistory extHistory =
+                Control.getSingleton().getExtensionLoader().getExtension(ExtensionHistory.class);
+        if (extHistory != null) {
+            EventQueue.invokeLater(
```
Comment (thc202) [inline addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarImporter.java:156]: This needs to be changed, if the view is not initialised it should not execute in the EDT.

File: addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarImporter.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+    private static HttpMessage getHttpMessage(HarEntry harEntry)
+            throws HttpMalformedHeaderException {
+        HttpMessage result = HarUtils.createHttpMessage(harEntry.getRequest());
+        setHttpResponse(harEntry.getResponse(), result);
+        setHistoryReference(harEntry, result);
+        return result;
+    }
+
+    private static void setHistoryReference(HarEntry harEntry, HttpMessage httpMessage) {
```
Comment (thc202) [inline addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarImporter.java:69]: I'd suggest removing this (and related logic in `persistMessage`), it's very unlikely that the IDs would still match.

**Comment by thc202** [comment]
I'm fine merging as is and do further changes in follow up PRs.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you both!

**Review by psiinon — APPROVED** [review]

**Pull request merged**